File: Java/Add-Two-Numbers.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;      
 *     }
 * }
 */
public class Solution {
    /**
     * @param l1: the first list
     * @param l2: the second list
     * @return: the sum list of l1 and l2 
     */
    public ListNode addLists(ListNode l1, ListNode l2) {
        // write your code here
        ListNode pre = new ListNode(-1), cur = pre;
        if (l1 == null) return l2;
        if (l2 == null) return l1;
        int carry = 0, num = 0;
        while (l1 != null && l2 != null) {
        	num = carry + l1.val + l2.val;
        	carry = num / 10;
        	num = num % 10;
        	l1.val = num;
        	cur.next = l1;
        	l1 = l1.next;
        	l2 = l2.next;
        	cur = cur.next;
        }
        while (l1 != null) {
        	num = carry + l1.val;
        	carry = num / 10;
        	num = num % 10;
        	l1.val = num;
        	cur.next = l1;
        	l1 = l1.next;
        	cur = cur.next;
        }
        while (l2 != null) {
        	num = carry + l2.val;
        	carry = num / 10;
        	num = num % 10;
        	l2.val = num;
        	cur.next = l2;
        	l2 = l2.next;
        	cur = cur.next;
        }
        if (carry != 0) {
        	ListNode temp = new ListNode(carry);
        	cur.next = temp;
        }
        return pre.next;
    }
}

File: Python/Add-Two-Numbers.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    # @param l1: the first list
    # @param l2: the second list
    # @return: the sum list of l1 and l2
    def helper(self, l1, l2, res):
        if l1 is None and l2 is None and res == 0:
            return None
        data = res
        if l1 is not None:
            data += l1.val
            l1 = l1.next
        if l2 is not None:
            data += l2.val
            l2 = l2.next
        node = ListNode(data % 10)
        node.next = self.helper(l1, l2, data / 10)
        return node

    def addLists(self, l1, l2):
        # write your code here
        return self.helper(l1, l2, 0)